raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.raft;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.ShortNode;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.raft.generated.QuorumStateData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.EOFException;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.util.OptionalInt;
+
+/**
+ * Local file based quorum state store. It takes the JSON format of {@link QuorumStateData}
+ * with an extra data version number as part of the data for easy deserialization.
+ *
+ * Example format:
+ * <pre>
+ * {"clusterId":"",
+ *   "leaderId":1,
+ *   "leaderEpoch":2,
+ *   "votedId":-1,
+ *   "appliedOffset":0,
+ *   "currentVoters":[],
+ *   "targetVoters":[],
+ *   "data_version":0}
+ * </pre>
+ * */
+public class FileBasedStateStore implements QuorumStateStore {
+    private static final Logger log = LoggerFactory.getLogger(FileBasedStateStore.class);
+
+    private final File stateFile;
+
+    private final QuorumStateData cachedState;
+
+    static final String DATA_VERSION = "data_version";
+
+    public FileBasedStateStore(final File stateFile, final ReplicatedLog metadataLog) throws IOException {
+        this.stateFile = stateFile;
+        cachedState = loadState(stateFile, metadataLog);
+        log.info("Loaded quorum state from local: {}", cachedState);
+    }
+
+    private QuorumStateData loadState(File file, final ReplicatedLog metadataLog) throws IOException {
```

Comment:
I think the problem of validating the quorum state against the epochs in the log is a common problem and not specific to the file based store. There is some initialization logic in `QuorumState.initialize` already:
```java
    public void initialize(long endOffset) throws IOException {
```

The end offset is taken from the log. I think we could change that signature to something like this instead:

```java
    public void initialize(OffsetAndEpoch endOffsetAndEpoch) throws IOException {
```
Whenever we initialize the quorum state, we can validate the epoch from the log. If the epoch from the log is larger, then we can write a new quorum state file. The advantage of doing it there is that we handle the case consistently for all `QuorumStateStore` implementations and we don't need the dependence on `ReplicatedLog` in this class.